src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -222,6 +262,73 @@ export function expected<T extends Dimensionality>(
   }
 }
 
+/**
+ * The software version of a texture builtin (eg: textureSample)
+ * Note that this is not a complete implementation. Rather it's only
+ * what's needed to generate the correct expected value for the tests.
+ */
+export function softwareTextureRead<T extends Dimensionality>(
+  call: TextureCall<T>,
+  texture: Texture,
+  sampler: GPUSamplerDescriptor,
+  size: [number, number]
+): PerTexelComponent<number> {
+  assert(call.ddx !== undefined);
+  assert(call.ddy !== undefined);
+  const rep = kTexelRepresentationInfo[texture.texels[0].format];
+  const texSize = reifyExtent3D(texture.descriptor.size);
+  const textureSize = [texSize.width, texSize.height];
+
+  const ddx: readonly number[] = typeof call.ddx === 'number' ? [call.ddx] : call.ddx;
+  const ddy: readonly number[] = typeof call.ddy === 'number' ? [call.ddy] : call.ddy;
+  const sDdx = ddx.map((v, i) => (v * textureSize[i]) / size[i]);
+  const sDdy = ddy.map((v, i) => (v * textureSize[i]) / size[i]);
```

Comment:
Comment what these variables are. Maybe stick to leibniz notation, rename ddx/ddy to dPdx/dPdy (define what P is) and sDdx/sDdy to dSdx/dSdy (define what S is)

Reply:
I changed the way this works. I changed this so `ddx` and `ddy` are the values passed to `textureSampleGrad` and for this test cases they're computed outside.